### PR TITLE
cmake: disable array-bounds check & treat warnings as errors for pyextobj and oneflow_internal & fix warnings

### PR DESCRIPTION
**Pull request by PragmaTwice**
- disable array-bounds check
- treat warnings as errors for pyextobj and oneflow_internal
- fix warnings
  - a null pointer dereference reported by gcc
  - some missing `final` to prevent missing dispatching in derived classes due to no virtual dtor
- tested in gcc 11 and clang 12
- check whether option is avaliable using `check_cxx_compiler_flag `

### CODE REVIEW

File: cmake/oneflow.cmake [hunk 1]
```diff
@@ -16,6 +16,32 @@ function(oneflow_add_library)
...
+      target_compile_options(${target} PRIVATE -Wno-error=unused-lambda-capture)
+
+      target_compile_options(${target} PRIVATE -Wno-error=instantiation-after-specialization)
+
+      # the mangled name between `struct X` and `class X` is different in MSVC ABI, remove it while windows is supported (in MSVC/cl or clang-cl)
+      target_compile_options(${target} PRIVATE -Wno-error=mismatched-tags)
+    endif()
+
+    # disable for pointer operations of intrusive linked lists
+    target_compile_options(${target} PRIVATE -Wno-error=array-bounds)
```
Comment (daquexian) [inline cmake/oneflow.cmake:41]: 写一个 target_compile_options_if_avaiable 宏用 check_cxx_compiler_flag 测试一下编译器对当前编译参数是不是支持，支持的时候再设置参数吧

Comment (PragmaTwice) [reply cmake/oneflow.cmake]: Done. afabe981442cb2ed95a63b404e71f3af3553ddf4

File: cmake/oneflow.cmake [hunk 2]
```diff
@@ -16,6 +17,59 @@ function(oneflow_add_library)
...
+  # We cannot check for -Wno-foo as this won't throw a warning so we must check for the -Wfoo option directly
+  # http://stackoverflow.com/questions/38785168/cc1plus-unrecognized-command-line-option-warning-on-any-other-warning
+  string(REGEX REPLACE "^-Wno-" "-W" checkedFlag ${flag})
+  string(REGEX REPLACE "[-=]" "_" varName CXX_FLAG${checkedFlag})
+  # Avoid double checks. A compiler will not magically support a flag it did not before
+  if(NOT ${varName}_CHECKED)
+    check_cxx_compiler_flag(${checkedFlag} ${varName}_SUPPORTED)
+    set(${varName}_CHECKED YES CACHE INTERNAL "")
+  endif()
+  if (${varName}_SUPPORTED)
```
Comment (daquexian) [inline cmake/oneflow.cmake:31]: 既然 ${varName}_SUPPORTED 也是 cache 变量，为什么还需要 ${varName}_CHECKED 呢

Comment (PragmaTwice) [reply cmake/oneflow.cmake:31]: _CHECKED 是指是否检查过，检查过之后 _SUPPORTED 可能是 false，所以如果没有 _CHECKED，遇到不合法的 option 时每次都要调用检查一遍

Comment (PragmaTwice) [reply cmake/oneflow.cmake:30]: Done

File: cmake/oneflow.cmake [hunk 3]
```diff
@@ -16,6 +17,58 @@ function(oneflow_add_library)
...
+  # We cannot check for -Wno-foo as this won't throw a warning so we must check for the -Wfoo option directly
+  # http://stackoverflow.com/questions/38785168/cc1plus-unrecognized-command-line-option-warning-on-any-other-warning
+  string(REGEX REPLACE "^-Wno-" "-W" checkedFlag ${flag})
+  string(REGEX REPLACE "[-=]" "_" varName CXX_FLAG${checkedFlag})
+  # Avoid double checks. A compiler will not magically support a flag it did not before
+  if(NOT DEFINED ${varName}_SUPPORTED)
+    check_cxx_compiler_flag(${checkedFlag} ${varName}_SUPPORTED)
+  endif()
+  if (${varName}_SUPPORTED)
+    message(STATUS "Add ${flag} to target ${target}")
```
Comment (daquexian) [inline cmake/oneflow.cmake:31]: 是不是在 supported 为 false 的时候也应该打印一个对应的日志

**Review by daquexian — APPROVED** [review]

**Pull request merged**